orttraining/orttraining/python/training/onnxblock/README.md

```diff
@@ -11,95 +11,102 @@ The ort training APIs need the following files for performing training
+1. export_params: True
+2. training: torch.onnx.TrainingMode.TRAINING
+3. do_constant_folding: False
+
+An example command for exporting the model can be:
+
+```py
+torch.onnx.export(model, sample_inputs, "base_model.onnx",
+                  export_params=True, training=torch.onnx.TrainingMode.TRAINING,
+                  do_constant_folding=False)
+```
+
+Now, onto generating the training artifacts. Let's assume that a forward only onnx model has already been generated. Here, we require that the user's onnx model is generated with the parameters embedded inside the exported model, i.e. with the parameters `export_params=True` and `training=torch.onnx.TrainingMode.TRAINING`.
+
+```py
+from onnxruntime.training import artifacts
+
+
+# Load the onnx model
+model_path = "model.onnx"
+base_model = onnx.load(model_path)
+
+# Define the parameters that need their gradient computed
+requires_grad = ["weight1", "bias1", "weight2", "bias2"]
+
+# Generate the training artifacts
+artifacts.generate_artifacts(base_model, requires_grad = requires_grad,
+                             loss = artifacts.LossType.CrossEntropyLoss,
+                             optimizer = artifacts.OptimType.AdamW)
+
+# Successful completion of the above call will generate 4 files in the current working directory,
+# one for each of the artifacts mentioned above (training_model.onnx, eval_model.onnx, checkpoint, optimizer_model.onnx)
+```
+
+Once the models and checkpoint have been generated, they can be loaded in the online training step and executed.
+For an example on how the online training loop should be written given these generated files, refer to this
+[sample trainer](https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/test/training_api/trainer/trainer.cc).
+
+### Advanced scenarios
+
+Let's assume, we want to use a custom loss function with a model. For this example, we assume that our model generates
+two outputs. And the custom loss function must apply a loss function on each of the outputs and perform a weighted average
+on the output. Mathematically,
+
+```
+loss = 0.4 * mse_loss1(output1, target1) + 0.6 * mse_loss2(output2, target2)
+```
+
+Since this is a custom loss function, this loss function is not exposed as an enum from `generate_artifacts` function.
+
+For this, we make use of `onnxblock`.
 
 ```py
 import onnxruntime.training.onnxblock as onnxblock
+from onnxruntime.training import artifacts
 
-# Define how the training model should be constructed.
-# In this example, the training model is constructed by stacking MSELoss
-# blocks on top of the inference model (with two outputs) and taking a
-# weighted average of the two losses to compute the final loss.
-# And the onnxblock tooling automatically takes care of building the
-# training graph for this model.
-class MyTrainingModel(onnxblock.TrainingModel):
+# Define a custom loss block that takes in two inputs
```

Comment:
Why not using torch to write the custom loss?

Reply:
This will require users to have torch in their offline tooling ecosystem.

While this could be true for many cases, it could also not be a dependency that the user would like to take.

For example, if the user begins with a model from model zoo, they would probably not like to take a dependency on torch for defining a custom loss.

But, as we spoke earlier, I want to keep this as a possibility. So, we may open up the input argument for custom loss to be a custom torch loss function. But I will prioritize this in a separate PR.